### ISSUE TITLE
Inaccurate `num_train_steps` in the LR scheduler during multi-GPU parallelism.

### ISSUE THREAD

**Issue by Dai-Wenxun** [huggingface/diffusers#6137]
### Describe the bug

When I was building my own LoRa training code, I encountered an issue with the `num_training_steps` parameter in the LR scheduler. I attempted to follow the official training script from [diffusers](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py#L787-L827). But I observed that the variable `args.max_train_steps` is not taking into account the division by the number of GPUs after the `accelerator.prepare` in the instantiation of the LR scheduler. This results in the `lr_scheduler.num_training_steps` being greater than the actual number of training steps. Consequently, this prevents adjustment strategies like cosine annealing from reaching 0. More precisely, the instantiation of the LR scheduler should derive `args.max_train_steps` from [here](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py#L825) then multiply the `accelerator.num_processes`. 

![d1ce6ae876d63c57b69f022d58c9a13](https://github.com/huggingface/diffusers/assets/58767402/74694aa5-0284-4a85-9838-d67cfdba5b89)

I use `get_cosine_with_hard_restarts_schedule_with_warmup` with `num_cycles=10`.

I found that the author of the consistency model seems to have identified this issue. Therefore, they handle the length of the dataloader for multiple GPUs before the `accelerator.prepare`. You can see the relevant code here: https://github.com/huggingface/diffusers/blob/main/examples/consistency_distillation/train_lcm_distill_sd_wds.py#L1025.

The attribute `train_dataloader.num_batches` takes into account the multi-GPU scenario during its calculation.

### Reproduction

    overrode_max_train_steps = False
    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / gradient_accumulation_steps)
    if max_train_steps == -1:
        max_train_steps = num_train_epochs * num_update_steps_per_epoch
        overrode_max_train_steps = True

    lr_scheduler = get_cosine_with_hard_restarts_schedule_with_warmup(
        optimizer, num_warmup_steps=lr_warmup_steps * accelerator.num_processes,
        num_training_steps=max_train_steps * accelerator.num_processes,
        num_cycles=lr_num_cycles
    )

    lora_net, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
        lora_net, optimizer, train_dataloader, lr_scheduler
    )

    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / gradient_accumulation_steps)
    if overrode_max_train_steps:
        max_train_steps = num_train_epochs * num_update_steps_per_epoch
    num_train_epochs = math.ceil(max_train_steps / num_update_steps_per_epoch)

### System Info

None

### Who can help?

 @sayakpaul @patrickvonplaten @yiyixuxu @patrickvonplaten @yiyixuxu @DN6 @sayakpaul @patrickvonplaten

**Comment by sayakpaul** [comment]
Feel free to propose a fix.

**Comment by Dai-Wenxun** [comment]
change [these lines](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py#L787-L827) to 
```python
    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / accelerator.num_processes / gradient_accumulation_steps)
    if args.max_train_steps is None:
        args.max_train_steps = args.num_train_epochs * num_update_steps_per_epoch

    lr_scheduler = get_scheduler(
        args.lr_scheduler,
        optimizer=optimizer,
        num_warmup_steps=args.lr_warmup_steps * accelerator.num_processes,
        num_training_steps=args.max_train_steps * accelerator.num_processes,
    )

    # Prepare everything with our `accelerator`.
    unet, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
        unet, optimizer, train_dataloader, lr_scheduler
    )

    if args.use_ema:
        ema_unet.to(accelerator.device)

    # For mixed precision training we cast all non-trainable weigths (vae, non-lora text_encoder and non-lora unet) to half-precision
    # as these weights are only used for inference, keeping weights in full precision is not required.
    weight_dtype = torch.float32
    if accelerator.mixed_precision == "fp16":
        weight_dtype = torch.float16
        args.mixed_precision = accelerator.mixed_precision
    elif accelerator.mixed_precision == "bf16":
        weight_dtype = torch.bfloat16
        args.mixed_precision = accelerator.mixed_precision

    # Move text_encode and vae to gpu and cast to weight_dtype
    text_encoder.to(accelerator.device, dtype=weight_dtype)
    vae.to(accelerator.device, dtype=weight_dtype)
```

The core modification is here since the `drop_last` defaults to `False`:
```python
num_update_steps_per_epoch = math.ceil(len(train_dataloader) / accelerator.num_processes / gradient_accumulation_steps)
```

**Comment by sayakpaul** [comment]
Great. I think it would be great if you opened a PR to contribute this :-)
